File: iosApp/iosApp/feature/resetPassword/ResetPasswordScreenViewModel.swift
//
//  ResetPasswordScreenViewModel.swift
//  iosApp
//
//  Created by Mikhail Maevskii on 13/11/2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import Shared

class ResetPasswordScreenViewModel: ObservableObject {
    var onNavigate: (Router.Destination) -> Void
    @Published var state = ResetPasswordScreenState()
    private let resetPasswordUseCase = ResetPasswordUseCaseWrapper().useCase
    
    init(onNavigate: @escaping (Router.Destination) -> Void) {
        self.onNavigate = onNavigate
    }
    
    func onEvent(event: ResetPasswordScreenEvent) {
        switch event {
        case .SendBtnClicked: onSendBtnClicked()
        case .MessageDialogDismissed: onMessageDialogDismissed()
        }
    }
    
    private func onSendBtnClicked() {
        if state.emailState.text.isEmpty {
            state = state.copy(screenState: ScreenState.result(resultType: ResultType.failure, messageCode: MessageCodes.fieldsCannotBeEmpty))
            return
        }
        state = state.copy(screenState: ScreenState.loading(messageCode: MessageCodes.sendingNewPassword))
        resetPassword(email: state.emailState.text)
    }
    private func onMessageDialogDismissed() {
        state = state.copy(screenState: ScreenState.idle)
    }
    
    private func resetPassword(email: String) {
        _Concurrency.Task {
            do {
                let res = try await resetPasswordUseCase.invoke(email: email)
                if res.isSuccess() {
                    await MainActor.run {
                        state = state.copy(screenState: ScreenState.result(
                            resultType: ResultType.success,
                            messageCode: MessageCodes.sentNewPassword
                        ))
                    }
                } else {
                    await MainActor.run {
                        state = state.copy(screenState: ScreenState.result(
                            resultType: ResultType.failure,
                            messageCode: MessageCodes.userWithEmailDoesntExist
                        ))
                    }
                }
            } catch {
                state = state.copy(screenState: ScreenState.result(
                    resultType: ResultType.failure,
                    messageCode: MessageCodes.gotUnexpectedErrorSendingNewPassword
                ))
            }
            
        }
    }
}

File: composeApp/src/androidMain/kotlin/dem/dev/timeflame/feature/resetPassword/ResetPasswordScreenViewModel.kt
package dem.dev.timeflame.feature.resetPassword

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dem.dev.timeflame.feature.resetPassword.usecase.ResetPasswordUseCase
import dem.dev.timeflame.util.state.ResultType
import dem.dev.timeflame.util.state.ScreenState
import dem.dev.timeflame.util.state.UiMessageCodes
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class ResetPasswordScreenViewModel(
    private val resetPasswordUseCase: ResetPasswordUseCase
): ViewModel() {
    private val _state = MutableStateFlow(ResetPasswordScreenState())
    val state = _state.asStateFlow()

    fun onEvent(event: ResetPasswordScreenEvent) {
        when(event) {
            is ResetPasswordScreenEvent.EmailChanged -> onEmailChanged(event.newEmail)
            ResetPasswordScreenEvent.SendNewPasswordBtnClicked -> onSendNewPasswordBtnClicked()
            ResetPasswordScreenEvent.MessageDialogDismissed -> onMessageDialogDismissed()
        }
    }

    private fun onEmailChanged(newEmail: String) {
        _state.update { it.copy(emailState = it.emailState.copy(value = newEmail)) }
    }
    private fun onSendNewPasswordBtnClicked() {
        _state.update { it.copy(screenState = ScreenState.Loading(UiMessageCodes.sendingNewPassword)) }
        sendNewPassword(email = _state.value.emailState.value)
    }
    private fun onMessageDialogDismissed() {
        _state.update { it.copy(screenState = ScreenState.Idle) }
    }

    private fun sendNewPassword(email: String) = viewModelScope.launch(Dispatchers.IO) {
        val res = resetPasswordUseCase(email)

        if (res.isSuccess())
            _state.update { it.copy(screenState = ScreenState.Result(ResultType.SUCCESS, UiMessageCodes.passwordWasSentToYourEmail)) }
        else
            _state.update { it.copy(screenState = ScreenState.Result(ResultType.FAILURE, UiMessageCodes.thereIsNoUserWithThisEmail))}
    }
}